plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+         return "(unknown)";
       }
    }
 
-   uint64_t get_log_entry_impl(const eosio::state_history::get_blocks_result& result,
-                               bool has_value,
-                               std::optional<state_history_log>& optional_log,
-                               std::optional<locked_decompress_stream>& buf) {
-      if (has_value) {
-         if( optional_log ) {
-            buf.emplace( optional_log->create_locked_decompress_stream() );
-            return std::visit([&](auto& r) { return optional_log->get_unpacked_entry( r.this_block->block_num, *buf ); }, result);
-         }
-      }
-      return 0;
-   }
-
-   uint64_t get_trace_log_entry(const eosio::state_history::get_blocks_result& result,
-                                std::optional<locked_decompress_stream>& buf) {
-      return std::visit([&](auto& r) { return get_log_entry_impl(r, r.traces.has_value(), plugin.get_trace_log(), buf); }, result);
-   }
-
-   uint64_t get_delta_log_entry(const eosio::state_history::get_blocks_result& result,
-                                std::optional<locked_decompress_stream>& buf) {
-      return std::visit([&](auto& r) { return get_log_entry_impl(r, r.deltas.has_value(), plugin.get_chain_state_log(), buf); }, result);
-   }
-
-   uint64_t get_finality_data_log_entry(const eosio::state_history::get_blocks_result_v1& result,
-                                        std::optional<locked_decompress_stream>& buf) {
-      return get_log_entry_impl(result, result.finality_data.has_value(), plugin.get_finality_data_log(), buf);
+   void awake_if_idle() {
+      boost::asio::dispatch(strand, [this]() {
+         wake_timer.cancel_one();
+      });
    }
 
-   void process(state_history::get_status_request_v0&) {
-      fc_dlog(plugin.get_logger(), "received get_status_request_v0");
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<status_result_send_queue_entry<session>>(self);
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
+   void check_coros_done(std::exception_ptr e) {
+      //the only exception that should have bubbled out of the coros is a bad_alloc, bubble it up further. No need to bother
+      //  with the rest of the cleanup: we'll be shutting down soon anyway due to bad_alloc
+      if(e)
+         std::rethrow_exception(e);
+      //coros always return on the session's strand
+      if(--coros_running == 0)
+         on_done(this);
    }
 
-   void process(state_history::get_blocks_request_v0& req) {
-      fc_dlog(plugin.get_logger(), "received get_blocks_request_v0 = ${req}", ("req", req));
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<blocks_request_send_queue_entry<session>>(self, std::move(req));
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
+   template<typename F>
+   void drop_exceptions(F&& f) {
+      try{ f(); } catch(...) {}
```

Comment:
That is correct, only time this is used is when performing final cleanup on an already failed connection. It's  dead, Jim. There isn't any reasonable action we could take besides logging, but logging could be deceptive since it wouldn't be logging why the connection failed -- just some other cleanup foul up. (the code goes through some trouble to always make sure it only logs one single error message on a connection failure, in hopes that the true culprit is logged)

I am not sure what specific conditions could cause these to throw. Both `close()` and `cancel_one()` are documented as possibly throwing a `boost::system::system_error` so I _have_ to do something. Letting anything escape here would, at the moment, land up in,
https://github.com/AntelopeIO/spring/blob/18bea833753b86c964c8723037666c40212853db/plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp#L76-L80
which would be a fatal ship thread error.

Reply:
The main risk I could see here is if somehow `close()` being called twice does something bad. I will look in to that.

Reply:
`close()`ing a socket multiple times isn't an error; I guess maybe that's not too surprising upon considering that when a socket is dtored it effectively gets `close()`ed too.

I'm not really sure what to do. Both `close()` and `cancel_one()` are documented as potentially throwing but the situations they throw might very much be corner cases or even pathological. Still we ought to be correct here and not blow the entire thread out. If having a "broad" `drop_exceptions()` is ugly maybe we could do

```c++
boost::system::error_code ignored;
stream.next_layer().close(ignored);
wake_timer.cancel_one(ignored); //HOWEVER, on steady_timer, this is deprecated usage 👎
```
I'd like to avoid depreciated interfaces, which leads us to something a little more icky like
```c++
boost::system::error_code ignored;
stream.next_layer().close(ignored);
try{ wake_timer.cancel_one();} catch(...) {} //ignore failures
```
this removes the need to have that broad `drop_exceptions()` hanging around. That